TCD1304.py

```diff
@@ -34,45 +32,56 @@
-        max_masterclock_frequency: int = 4e6,
-        frequency: float = 2e6, #Standart Frequenz der Masterclock sind 2 MHz
-        min_volatagerange_clock: int = 0,
-        max_volatagerange_clock: int = 4,
+        self, #Initializing; self is always first parameter of methods located inside a class and is not limited to the __init__ method but to most methods. A method is a function that belongs to a class
+        min_masterclock_frequency = 0.8e6, # no need to specify the variable type (float).  View comment in L37
+        max_masterclock_frequency = 4e6, # no need to specify the variable type (float). View comment in L37
+        frequency = frequency_master_clock, #default frequency of masterclock 2 MHz; # no need to specify the variable type (float). View comment in L37
+        min_voltagerange_clock: float = 0, #it says that min_voltagerange_clock should (!) be a float. Otherwise it would not turn into a float automatically. Output on float will be 0.0 V and on int 0 V.
+        max_voltagerange_clock: float = 4,
+        pwmgen = ... # not sure if pwmgen needs to be assigned to a value ??
         ): 
 
-        self.min_masterclock_frequency = min_masterclock_frequency #Hier wird initialisiert, wenn mich nicht alles täsucht
+        self.min_masterclock_frequency = min_masterclock_frequency #initializing starts from here (? Not sure though)
         self.max_masterclock_frequency = max_masterclock_frequency
         self.frequency = frequency
-        self.min_volatagerange_clock = min_volatagerange_clock
-        self.max_volatagerange_clock = max_volatagerange_clock
-
+        self.min_voltagerange_clock = min_voltagerange_clock
+        self.max_voltagerange_clock = max_voltagerange_clock
+        self.pwmgen = PWMGenerator() #once assigned in __init__ self.pwmgen can be used in other methods
+        
     def power_source ():        # puts a voltage of 4V on PV1
         ps = PowerSupply()
         ps.pv1 = 4
         ps.pv1
     
-    def master_clock ():        # puts PWM with frequency of 2MHz on SQ1
-        pwmgen = PWMGenerator()
-        pwmgen.generate(["SQ1"], fmc, [0.5])       
+    def master_clock (self):        # puts PWM with frequency of 2MHz on SQ1; masterclock is the fastest clock
+        prescaler = int(math.log(oscillator_frequency / frequency_master_clock) / math.log(2))   # When setting a frequency by mapping the reference clock directly to a PWM output, only frequencies which are even factors of 128 MHz (the frequency of the PSLab's main oscillator) are available. The frequency is therefore not set by specifying the frequency itself, but by setting a prescaler.
+        self.pwmgen.map_reference_clock(["SQ1"], prescaler)    
     
-    def sh_clock ():            #  on SQ2 sets pulse for the integrationtime
+    def sh_clock (self):           #  on SQ2 sets pulse for the integrationtime
+        return self.pwmgen.generate() #is the return attribute the one to choose in order to set a second clock?

```

Comment:
Here you need to set the frequency of the SH clock, check https://tcd1304.wordpress.com/timing-requirements/. This method should not return anything.